src/deriver/tom/embeddings.py

```
@@ -1,117 +1,706 @@
+                            f"Skipping duplicate observation during retrieval: {doc.content[:50]}..."
+                        )
+
+            return context
+
+        except Exception as e:
+            logger.error(f"Error retrieving relevant observations for reasoning: {e}")
+            # Fallback to recent observations if semantic search fails
+            logger.warning("Falling back to recent observations retrieval")
+            return await self.get_most_recent_observations()
+
+    async def execute_queries(self, queries: list[str]) -> "QueryExecutionResult":
```

Ruff (0.11.9)

Undefined name `QueryExecutionResult`

(F821)